File: ChunkL/Lexing/Lexer.cs
namespace ChunkL.Lexing;

/// <summary>
/// Tokenizes a ChunkL source string into a list of tokens.
/// </summary>
public sealed class Lexer
{
    private readonly string _source;
    private int _pos;
    private int _line = 1;
    private int _column = 1;
    private int _leadingSpaces;
    private bool _atLineStart = true;

    public Lexer(string source)
    {
        _source = source;
    }

    public List<Token> Tokenize()
    {
        var tokens = new List<Token>();

        while (_pos < _source.Length)
        {
            var ch = _source[_pos];

            // Track leading whitespace at start of line
            if (_atLineStart && ch == ' ')
            {
                _leadingSpaces = 0;
                while (_pos < _source.Length && _source[_pos] == ' ')
                {
                    _leadingSpaces++;
                    Advance();
                }
                _atLineStart = false;
                continue;
            }

            if (_atLineStart)
            {
                _leadingSpaces = 0;
                _atLineStart = false;
            }

            // Skip non-leading whitespace (spaces/tabs within a line)
            if (ch == ' ' || ch == '\t')
            {
                Advance();
                continue;
            }

            // Newlines
            if (ch == '\n')
            {
                tokens.Add(MakeToken(TokenKind.Newline, "\n", _pos, 1));
                Advance();
                _atLineStart = true;
                continue;
            }

            if (ch == '\r')
            {
                if (_pos + 1 < _source.Length && _source[_pos + 1] == '\n')
                {
                    tokens.Add(MakeToken(TokenKind.Newline, "\r\n", _pos, 2));
                    Advance();
                    Advance();
                }
                else
                {
                    tokens.Add(MakeToken(TokenKind.Newline, "\r", _pos, 1));
                    Advance();
                }
                _atLineStart = true;
                continue;
            }

            // Comments: // or #
            if (ch == '/' && _pos + 1 < _source.Length && _source[_pos + 1] == '/')
            {
                var start = _pos;
                var startCol = _column;
                var startLine = _line;
                while (_pos < _source.Length && _source[_pos] != '\n' && _source[_pos] != '\r')
                {
                    Advance();
                }
                var text = _source.Substring(start, _pos - start);
                tokens.Add(new Token(TokenKind.Comment, text, new SourcePosition(startLine, startCol), _leadingSpaces, start, _pos - start));
                continue;
            }

            if (ch == '#')
            {
                var start = _pos;
                var startCol = _column;
                var startLine = _line;
                while (_pos < _source.Length && _source[_pos] != '\n' && _source[_pos] != '\r')
                {
                    Advance();
                }
                var text = _source.Substring(start, _pos - start);
                tokens.Add(new Token(TokenKind.Comment, text, new SourcePosition(startLine, startCol), _leadingSpaces, start, _pos - start));
                continue;
            }

            // String literals
            if (ch == '"')
            {
                tokens.Add(ReadStringLiteral());
                continue;
            }

            // Hex literals: 0x...
            if (ch == '0' && _pos + 1 < _source.Length && (_source[_pos + 1] == 'x' || _source[_pos + 1] == 'X'))
            {
                tokens.Add(ReadHexLiteral());
                continue;
            }

            // Integer literals (or negative if preceded by context)
            if (char.IsDigit(ch))
            {
                tokens.Add(ReadIntLiteral());
                continue;
            }

            // Identifiers and keywords
            if (IsIdentStart(ch))
            {
                tokens.Add(ReadIdentifier());
                continue;
            }

            // Multi-char operators
            tokens.Add(ReadOperator());
        }

        tokens.Add(MakeToken(TokenKind.EndOfFile, "", _pos, 0));
        return tokens;
    }

    private Token ReadStringLiteral()
    {
        var start = _pos;
        var startCol = _column;
        var startLine = _line;

        Advance(); // skip opening "
        while (_pos < _source.Length && _source[_pos] != '"' && _source[_pos] != '\n' && _source[_pos] != '\r')
        {
            if (_source[_pos] == '\\' && _pos + 1 < _source.Length)
            {
                Advance(); // skip escape char
            }
            Advance();
        }

        if (_pos < _source.Length && _source[_pos] == '"')
        {
            Advance(); // skip closing "
        }

        var text = _source.Substring(start, _pos - start);
        return new Token(TokenKind.StringLiteral, text, new SourcePosition(startLine, startCol), _leadingSpaces, start, _pos - start);
    }

    private Token ReadHexLiteral()
    {
        var start = _pos;
        var startCol = _column;
        var startLine = _line;

        Advance(); // 0
        Advance(); // x

        while (_pos < _source.Length && IsHexDigit(_source[_pos]))
        {
            Advance();
        }

        var text = _source.Substring(start, _pos - start);
        return new Token(TokenKind.HexLiteral, text, new SourcePosition(startLine, startCol), _leadingSpaces, start, _pos - start);
    }

    private Token ReadIntLiteral()
    {
        var start = _pos;
        var startCol = _column;
        var startLine = _line;

        while (_pos < _source.Length && char.IsDigit(_source[_pos]))
        {
            Advance();
        }

        // Check for float-like suffix (e.g. 0.5f or 1.0)
        if (_pos < _source.Length && _source[_pos] == '.' && _pos + 1 < _source.Length && char.IsDigit(_source[_pos + 1]))
        {
            Advance(); // skip .
            while (_pos < _source.Length && char.IsDigit(_source[_pos]))
            {
                Advance();
            }
            // Optional f suffix
            if (_pos < _source.Length && (_source[_pos] == 'f' || _source[_pos] == 'F'))
            {
                Advance();
            }
        }
        else if (_pos < _source.Length && (_source[_pos] == 'f' || _source[_pos] == 'F'))
        {
            Advance();
        }

        var text = _source.Substring(start, _pos - start);
        return new Token(TokenKind.IntLiteral, text, new SourcePosition(startLine, startCol), _leadingSpaces, start, _pos - start);
    }

    private Token ReadIdentifier()
    {
        var start = _pos;
        var startCol = _column;
        var startLine = _line;

        while (_pos < _source.Length && IsIdentPart(_source[_pos]))
        {
            Advance();
        }

        var text = _source.Substring(start, _pos - start);
        return new Token(TokenKind.Identifier, text, new SourcePosition(startLine, startCol), _leadingSpaces, start, _pos - start);
    }

    private Token ReadOperator()
    {
        var ch = _source[_pos];
        var start = _pos;

        switch (ch)
        {
            case '.':
                if (_pos + 1 < _source.Length && _source[_pos + 1] == '.')
                {
                    Advance(); Advance();
                    return new Token(TokenKind.DotDot, "..", new SourcePosition(_line, _column - 2), _leadingSpaces, start, 2);
                }
                Advance();
                return new Token(TokenKind.Dot, ".", new SourcePosition(_line, _column - 1), _leadingSpaces, start, 1);

            case ',':
                Advance();
                return new Token(TokenKind.Comma, ",", new SourcePosition(_line, _column - 1), _leadingSpaces, start, 1);

            case ':':
                if (_pos + 1 < _source.Length && _source[_pos + 1] == ':')
                {
                    Advance(); Advance();
                    return new Token(TokenKind.ColonColon, "::", new SourcePosition(_line, _column - 2), _leadingSpaces, start, 2);
                }
                Advance();
                return new Token(TokenKind.Colon, ":", new SourcePosition(_line, _column - 1), _leadingSpaces, start, 1);

            case '=':
                if (_pos + 1 < _source.Length && _source[_pos + 1] == '=')
                {
                    Advance(); Advance();
                    return new Token(TokenKind.EqualsEquals, "==", new SourcePosition(_line, _column - 2), _leadingSpaces, start, 2);
                }
                Advance();
                return new Token(TokenKind.Equals, "=", new SourcePosition(_line, _column - 1), _leadingSpaces, start, 1);

            case '!':
                if (_pos + 1 < _source.Length && _source[_pos + 1] == '=')
                {
                    Advance(); Advance();
                    return new Token(TokenKind.BangEquals, "!=", new SourcePosition(_line, _column - 2), _leadingSpaces, start, 2);
                }
                Advance();
                return new Token(TokenKind.Bang, "!", new SourcePosition(_line, _column - 1), _leadingSpaces, start, 1);

            case '<':
                if (_pos + 1 < _source.Length && _source[_pos + 1] == '=')
                {
                    Advance(); Advance();
                    return new Token(TokenKind.LessThanEquals, "<=", new SourcePosition(_line, _column - 2), _leadingSpaces, start, 2);
                }
                if (_pos + 1 < _source.Length && _source[_pos + 1] == '<')
                {
                    Advance(); Advance();
                    return new Token(TokenKind.LessLess, "<<", new SourcePosition(_line, _column - 2), _leadingSpaces, start, 2);
                }
                Advance();
                return new Token(TokenKind.LessThan, "<", new SourcePosition(_line, _column - 1), _leadingSpaces, start, 1);

            case '>':
                if (_pos + 1 < _source.Length && _source[_pos + 1] == '=')
                {
                    Advance(); Advance();
                    return new Token(TokenKind.GreaterThanEquals, ">=", new SourcePosition(_line, _column - 2), _leadingSpaces, start, 2);
                }
                if (_pos + 1 < _source.Length && _source[_pos + 1] == '>')
                {
                    Advance(); Advance();
                    return new Token(TokenKind.GreaterGreater, ">>", new SourcePosition(_line, _column - 2), _leadingSpaces, start, 2);
                }
                Advance();
                return new Token(TokenKind.GreaterThan, ">", new SourcePosition(_line, _column - 1), _leadingSpaces, start, 1);

            case '(':
                Advance();
                return new Token(TokenKind.OpenParen, "(", new SourcePosition(_line, _column - 1), _leadingSpaces, start, 1);

            case ')':
                Advance();
                return new Token(TokenKind.CloseParen, ")", new SourcePosition(_line, _column - 1), _leadingSpaces, start, 1);

            case '[':
                Advance();
                return new Token(TokenKind.OpenBracket, "[", new SourcePosition(_line, _column - 1), _leadingSpaces, start, 1);

            case ']':
                Advance();
                return new Token(TokenKind.CloseBracket, "]", new SourcePosition(_line, _column - 1), _leadingSpaces, start, 1);

            case '+':
                Advance();
                return new Token(TokenKind.Plus, "+", new SourcePosition(_line, _column - 1), _leadingSpaces, start, 1);

            case '-':
                Advance();
                return new Token(TokenKind.Minus, "-", new SourcePosition(_line, _column - 1), _leadingSpaces, start, 1);

            case '*':
                Advance();
                return new Token(TokenKind.Star, "*", new SourcePosition(_line, _column - 1), _leadingSpaces, start, 1);

            case '/':
                Advance();
                return new Token(TokenKind.Slash, "/", new SourcePosition(_line, _column - 1), _leadingSpaces, start, 1);

            case '&':
                if (_pos + 1 < _source.Length && _source[_pos + 1] == '&')
                {
                    Advance(); Advance();
                    return new Token(TokenKind.AmpersandAmpersand, "&&", new SourcePosition(_line, _column - 2), _leadingSpaces, start, 2);
                }
                Advance();
                return new Token(TokenKind.Ampersand, "&", new SourcePosition(_line, _column - 1), _leadingSpaces, start, 1);

            case '|':
                if (_pos + 1 < _source.Length && _source[_pos + 1] == '|')
                {
                    Advance(); Advance();
                    return new Token(TokenKind.PipePipe, "||", new SourcePosition(_line, _column - 2), _leadingSpaces, start, 2);
                }
                Advance();
                return new Token(TokenKind.Pipe, "|", new SourcePosition(_line, _column - 1), _leadingSpaces, start, 1);

            case '~':
                Advance();
                return new Token(TokenKind.Tilde, "~", new SourcePosition(_line, _column - 1), _leadingSpaces, start, 1);

            case '^':
                Advance();
                return new Token(TokenKind.Caret, "^", new SourcePosition(_line, _column - 1), _leadingSpaces, start, 1);

            case '?':
                Advance();
                return new Token(TokenKind.Question, "?", new SourcePosition(_line, _column - 1), _leadingSpaces, start, 1);

            default:
                Advance();
                return new Token(TokenKind.Unknown, ch.ToString(), new SourcePosition(_line, _column - 1), _leadingSpaces, start, 1);
        }
    }

    private Token MakeToken(TokenKind kind, string text, int offset, int length)
    {
        return new Token(kind, text, new SourcePosition(_line, _column), _leadingSpaces, offset, length);
    }

    private void Advance()
    {
        if (_pos < _source.Length)
        {
            if (_source[_pos] == '\n')
            {
                _line++;
                _column = 1;
            }
            else
            {
                _column++;
            }
            _pos++;
        }
    }

    private static bool IsIdentStart(char ch) => char.IsLetter(ch) || ch == '_';
    private static bool IsIdentPart(char ch) => char.IsLetterOrDigit(ch) || ch == '_';
    private static bool IsHexDigit(char ch) =>
        char.IsDigit(ch) || (ch >= 'a' && ch <= 'f') || (ch >= 'A' && ch <= 'F');
}

File: chunkl-rs/src/lexer.rs
use crate::SourcePosition;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TokenKind {
    HexLiteral,
    IntLiteral,
    Identifier,
    StringLiteral,
    Dot,
    DotDot,
    Comma,
    Colon,
    ColonColon,
    Equals,
    EqualsEquals,
    BangEquals,
    LessThan,
    GreaterThan,
    LessThanEquals,
    GreaterThanEquals,
    LessLess,
    GreaterGreater,
    OpenParen,
    CloseParen,
    OpenBracket,
    CloseBracket,
    Plus,
    Minus,
    Star,
    Slash,
    Ampersand,
    AmpersandAmpersand,
    Pipe,
    PipePipe,
    Bang,
    Tilde,
    Caret,
    Question,
    Comment,
    Newline,
    EndOfFile,
    Unknown,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Token {
    pub kind: TokenKind,
    pub text: String,
    pub position: SourcePosition,
    pub leading_spaces: usize,
    pub source_offset: usize,
    pub source_length: usize,
}

pub struct Lexer<'a> {
    source: &'a str,
    offset: usize,
    line: usize,
    column: usize,
    leading_spaces: usize,
    at_line_start: bool,
}

impl<'a> Lexer<'a> {
    pub fn new(source: &'a str) -> Self {
        Self {
            source,
            offset: 0,
            line: 1,
            column: 1,
            leading_spaces: 0,
            at_line_start: true,
        }
    }

    pub fn tokenize(mut self) -> Vec<Token> {
        let mut tokens = Vec::new();
        while self.offset < self.source.len() {
            self.skip_horizontal_whitespace();
            if self.offset >= self.source.len() {
                break;
            }

            let start = self.offset;
            let position = SourcePosition::new(self.line, self.column);
            let indent = self.leading_spaces;
            let byte = self.current_byte();

            let (kind, end) = if byte == b'\r' || byte == b'\n' {
                let end = self.consume_newline();
                (TokenKind::Newline, end)
            } else if self.rest().starts_with("//") || byte == b'#' {
                (TokenKind::Comment, self.scan_to_newline())
            } else if byte == b'"' {
                (TokenKind::StringLiteral, self.scan_string())
            } else if self.rest().starts_with("0x") || self.rest().starts_with("0X") {
                (TokenKind::HexLiteral, self.scan_hex())
            } else if byte.is_ascii_digit() {
                (TokenKind::IntLiteral, self.scan_number())
            } else if is_identifier_start(byte) {
                (TokenKind::Identifier, self.scan_identifier())
            } else if let Some((kind, length)) = multi_char_token(self.rest()) {
                self.advance_bytes(length);
                (kind, self.offset)
            } else {
                let kind = single_char_token(byte).unwrap_or(TokenKind::Unknown);
                self.advance_bytes(1);
                (kind, self.offset)
            };

            tokens.push(Token {
                kind,
                text: self.source[start..end].to_owned(),
                position,
                leading_spaces: indent,
                source_offset: start,
                source_length: end - start,
            });
            if kind != TokenKind::Newline {
                self.at_line_start = false;
            }
        }

        tokens.push(Token {
            kind: TokenKind::EndOfFile,
            text: String::new(),
            position: SourcePosition::new(self.line, self.column),
            leading_spaces: self.leading_spaces,
            source_offset: self.offset,
            source_length: 0,
        });
        tokens
    }

    fn rest(&self) -> &str {
        &self.source[self.offset..]
    }

    fn current_byte(&self) -> u8 {
        self.source.as_bytes()[self.offset]
    }

    fn skip_horizontal_whitespace(&mut self) {
        while self.offset < self.source.len() {
            match self.current_byte() {
                b' ' => {
                    if self.at_line_start {
                        self.leading_spaces += 1;
                    }
                    self.advance_bytes(1);
                }
                b'\t' => {
                    if self.at_line_start {
                        self.leading_spaces += 4;
                    }
                    self.advance_bytes(1);
                }
                _ => break,
            }
        }
    }

    fn advance_bytes(&mut self, count: usize) {
        self.offset += count;
        self.column += count;
    }

    fn consume_newline(&mut self) -> usize {
        if self.rest().starts_with("\r\n") {
            self.offset += 2;
        } else {
            self.offset += 1;
        }
        self.line += 1;
        self.column = 1;
        self.leading_spaces = 0;
        self.at_line_start = true;
        self.offset
    }

    fn scan_to_newline(&mut self) -> usize {
        while self.offset < self.source.len() && !matches!(self.current_byte(), b'\r' | b'\n') {
            self.advance_bytes(1);
        }
        self.offset
    }

    fn scan_string(&mut self) -> usize {
        self.advance_bytes(1);
        let mut escaped = false;
        while self.offset < self.source.len() {
            let byte = self.current_byte();
            if matches!(byte, b'\r' | b'\n') {
                break;
            }
            self.advance_bytes(1);
            if byte == b'"' && !escaped {
                break;
            }
            escaped = byte == b'\\' && !escaped;
            if byte != b'\\' {
                escaped = false;
            }
        }
        self.offset
    }

    fn scan_hex(&mut self) -> usize {
        self.advance_bytes(2);
        while self.offset < self.source.len() && self.current_byte().is_ascii_hexdigit() {
            self.advance_bytes(1);
        }
        self.offset
    }

    fn scan_number(&mut self) -> usize {
        while self.offset < self.source.len()
            && (self.current_byte().is_ascii_digit()
                || matches!(self.current_byte(), b'.' | b'f' | b'F'))
        {
            if self.rest().starts_with("..") {
                break;
            }
            self.advance_bytes(1);
        }
        self.offset
    }

    fn scan_identifier(&mut self) -> usize {
        while self.offset < self.source.len() && is_identifier_continue(self.current_byte()) {
            self.advance_bytes(1);
        }
        self.offset
    }
}

fn is_identifier_start(byte: u8) -> bool {
    byte.is_ascii_alphabetic() || byte == b'_'
}

fn is_identifier_continue(byte: u8) -> bool {
    byte.is_ascii_alphanumeric() || byte == b'_'
}

fn multi_char_token(source: &str) -> Option<(TokenKind, usize)> {
    [
        ("==", TokenKind::EqualsEquals),
        ("!=", TokenKind::BangEquals),
        ("<=", TokenKind::LessThanEquals),
        (">=", TokenKind::GreaterThanEquals),
        ("<<", TokenKind::LessLess),
        (">>", TokenKind::GreaterGreater),
        ("::", TokenKind::ColonColon),
        ("..", TokenKind::DotDot),
        ("&&", TokenKind::AmpersandAmpersand),
        ("||", TokenKind::PipePipe),
    ]
    .into_iter()
    .find_map(|(text, kind)| source.starts_with(text).then_some((kind, text.len())))
}

fn single_char_token(byte: u8) -> Option<TokenKind> {
    Some(match byte {
        b'.' => TokenKind::Dot,
        b',' => TokenKind::Comma,
        b':' => TokenKind::Colon,
        b'=' => TokenKind::Equals,
        b'<' => TokenKind::LessThan,
        b'>' => TokenKind::GreaterThan,
        b'(' => TokenKind::OpenParen,
        b')' => TokenKind::CloseParen,
        b'[' => TokenKind::OpenBracket,
        b']' => TokenKind::CloseBracket,
        b'+' => TokenKind::Plus,
        b'-' => TokenKind::Minus,
        b'*' => TokenKind::Star,
        b'/' => TokenKind::Slash,
        b'&' => TokenKind::Ampersand,
        b'|' => TokenKind::Pipe,
        b'!' => TokenKind::Bang,
        b'~' => TokenKind::Tilde,
        b'^' => TokenKind::Caret,
        b'?' => TokenKind::Question,
        _ => return None,
    })
}
